### PR TITLE
4.0.0 - Simpler lambdas and the death of Template Haskell

### PR DESCRIPTION
**Pull request by NickSeagull**
After the discussion at #78 , it is clear that the way to go is to have a simpler `runLambda` function that users can use in order to make better decisions regarding their applications.

Apart from that, with the new [additions for AWS SAM](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/building-custom-runtimes.html) it makes even less sense to have a dynamic `dispatch` function. This autogenerated function not only was horrible for the users to debug, but the Template Haskell code that generated this function was horrible to maintain, and test, as well.

The roadmap for the 4.0.0 version is:

- [x] Remove the Template Haskell code
- [x] Fix `runLambda` so it is easier to call from a `main` function in an executable by the user
- [ ] Provide a better way to wrap the user code, so it receives different event types from different services, e.g. API GateWay
- [ ] Initialize `amazonka` on cold starts, so users can write their code either in the `AWST` monad, `IO` or anything in between
- [ ] Document the new version, including how to use it with AWS SAM

Inviting @dnikolovv @lrworth @endgame for discussing this.

### CODE REVIEW

**Comment by dnikolovv** [comment]
Sounds pretty awesome to me! Let me know if I can help with anything.

**Comment by endgame** [comment]
Will look at the code a bit later, but some pointers to get you started on the documentation:

* AWS SAM page on building custom runtimes: https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/building-custom-runtimes.html
* Each `AWS::Serverless::Function` (e.g. `Foo`) declared with `Metadata: { BuildMethod: Makefile }` in `template.yaml` (or `.json`) will cause `sam` to run `make build-Foo` in the directory specified by that function's `CodeUri` property.
* Each target is responsible for populating `$(ARTIFACTS_DIR)`, an environment variable set by `sam` when it triggers `make`. For most haskell projects, building one executable target and copying it to `$(ARTIFACTS_DIR)/bootstrap` seems to be sufficient.

**Comment by lrworth** [comment]
Wow, thanks @NickSeagull — this sounds fantastic! I'll also have a look later. I'm more than happy to contribute any work that needs doing and test stuff, just let me know.

(Also it looks like you're using ormolu for formatting, excellent choice IMO)

File: src/Aws/Lambda/Runtime.hs [hunk 1]
```diff
@@ -29,65 +20,103 @@ import qualified Aws.Lambda.Runtime.Environment as Environment
 import qualified Aws.Lambda.Runtime.Error as Error
 import qualified Aws.Lambda.Runtime.Publish as Publish
 import qualified Control.Exception as Unchecked
+import Control.Exception.Safe.Checked
```
Comment (endgame) [inline src/Aws/Lambda/Runtime.hs:23]: `safe-exceptions-checked` is tagged `deprecated` on Hackage.

File: src/Aws/Lambda/Runtime.hs [hunk 2]
```diff
@@ -29,65 +20,103 @@ import qualified Aws.Lambda.Runtime.Environment as Environment
...
-
+    event <- ApiInfo.fetchEvent manager lambdaApi `catch` errorParsing
     -- Purposefully shadowing to prevent using the initial "empty" context
     context <- Context.setEventData context event
-
-    (((invokeAndRun callback manager lambdaApi event context
-      `Checked.catch` \err -> Publish.parsingError err lambdaApi context manager)
-      `Checked.catch` \err -> Publish.invocationError err lambdaApi context manager)
-      `Checked.catch` \(err :: Error.EnvironmentVariableNotSet) -> Publish.runtimeInitError err lambdaApi context manager)
+    ( ( ( invokeAndRun callback manager lambdaApi event context
```
Comment (endgame) [inline src/Aws/Lambda/Runtime.hs:74]: I am not sure if you are forced to use `safe-exceptions-checked` by one of your dependencies, but if you are not, `catches` in `Control.Exception` may be tidier here: https://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Exception.html#v:catches

**Comment by NickSeagull** [comment]
Thanks everyone for the comments, I'm gonna try pushing forward a minimal version of this for 4.0.0, I'd need some help with writing unit tests and integration too.

**Comment by endgame** [comment]
Something we have discovered w.r.t. the custom runtime support: `sam` copies the source (i.e., the `CodeUri`) directory into a temporary location to do this build. It does this to maintain a pristine source tree.

This has some implications that should be documented. Our needs are a bit more complicated than the norm - we have a single repo with multiple services and multiple libraries; simpler setups can probably put `template.yaml` and `Makefile` next to the top-level `cabal.project` or `stack.yaml`. I'll explain them below so other people don't fall into the same holes we did:

1. Builds that use relative paths pointing outside the build directory won't work - see https://github.com/awslabs/aws-sam-cli/issues/2077 (which contains a sample directory tree by way of example). There are a few ways to deal with this:
  * Symlink the build directories so the symlink gets copied along with the source.
  * Make sure `CodeUri` points to a common ancestor directory of any libraries referenced locally. This does mean that you may copy a lot of irrelevant stuff to the build directory, and you may have to put the `Makefile` in the same directory. This also means that you may have a bunch of different project targets in the same `Makefile`.
  * Use tricks to find the real source directory, and fix up references (e.g., `sed`ing `cabal.project` or `stack.yaml`) before building. This workaround is described in the linked issue.
2. If your build tool caches its output under the source directory, you lose a lot of that when the tempdir goes away (i.e., after each target it built), meaning that each build target rebuilds all the local libraries, as well as common code. This is harder to fix, and **does** impact people with less complicated use-cases, if they're reusing code in a `library` stanza:
  * Use a build tool which caches elsewhere, so you can get reuse the local libraries that you build (e.g., use nix to cache your local library dependencies?); and
  * Make code that is shared between endpoints a separate cabal package, so that it can get this sharing too. Something like a `servicename-core` containing the library/tests, and a `servicename-endpoints` containing the code for each executable.

We avoided \#2 by using shell scripts to prepare our deployment packages. This means we still get the benefit of simpler and smaller endpoint executables.

Sorry, that was a bit of a braindump; I hope that it helps. LMK if anything's unclear and I'll have another go at explaining.

**Comment by dnikolovv** [comment]
@endgame We have solved the relative build paths issue by building using Docker and passing in the root of the repo as the context.

Our Makefiles look something like this.

```make
all:
	@rm -rf ./build/*
	DOCKER_BUILDKIT=1 docker build --file Dockerfile -t service ../../.
	id=$$(docker create service); docker cp $$id:/root/output ./build; docker rm -v $$id
```

**Comment by NickSeagull** [comment]
Hey, just posting a message to say that I haven't abandoned this, just that I didn't have time due to life reasons. Will soon resume work on this :)

**Comment by talw** [comment]
I wonder if this is still in progress?

**Comment by NickSeagull** [comment]
This is pretty much stalled right now, would be happy to help if someone wants to pick this up

**Comment by talw** [comment]
> This is pretty much stalled right now, would be happy to help if someone wants to pick this up

I see.
Unfortunately my hands are full at the moment as well.

Hopefully someone will pick this up :)

**Comment by fcracker79** [comment]
Hi there,

I am trying to implement multi-handler lambda and the template expansion does not allow it, as some modules are not exposed.
Will this fix solve my use case?

In addition to that, while waiting for this pull request to be merged, I could work on a simpler change that would expose {{ApiGatewayLambdaError}}. Does it make sense?

**Comment by NickSeagull** [comment]
@fcracker79 can you open an issue for each of those topics so we can discuss over there?

**Comment by fcracker79** [comment]
@NickSeagull thanks for your response.
There is already an issue related to my question (https://github.com/theam/aws-lambda-haskell-runtime/issues/82#issuecomment-659753726). I was just asking if this pull request will solve that issue. 

In addition to that, since I would like to proceed with my project and this pull request has been WIP for weeks, I was thinking of creating a new pull request with a simple change to export `ApiGatewayLambdaError`; I just need a piece of advice: wait for the completion of this pull request or create a far simpler one that temporarily solves the above issue but will be deprecated by this pull request?

**Comment by NickSeagull** [comment]
@fcracker79 I'm not actively working on this at this moment, so your PR would be well accepted. Regarding to #82 note that the issue is that the handler generator is hardcoded to look for the string `handler ::` in the project files, as the issue OP solved

**Comment by fcracker79** [comment]
Thanks @NickSeagull , currently it would be more than enough for my project to be able to copy and paste that generated code and then change all the hardcoded strings I want so that to implement multi handler entry point.

**Comment by NickSeagull** [comment]
Closing as #97 did this and more in a better way :)

**Pull request closed without merging**